GCS connector requires storage.buckets.get permissions to list entities at the bucket root

The `StorageObjectAdmin` Role is not sufficient to able to use the GCS connector to perform the following operations
1) ls at the root ->` hadoop fs -ls gs://BUCKET/`
2) ls on an objet that does not exist -> `hadoop fs -ls gs://BUCKET/{INVALID_OBJECT}`

The exception is pasted further down.

The same is not the case for `gsutil`.
`gsutil ls gs://BUCKET/` and `gsutil ls gs://BUCKET/{INVALID_OBJECT}` works fine.

It looks like the connector falls back to perform a `bucket.get` operation. Creating a custom role with `storage.bucket.get` + `StorageObjectAdmin` gets the connector working properly again.

Is this intended behaviour? If so, would be nice to document.

Here's the stack trace.

```
Exception in thread "main" java.io.IOException: Multiple IOExceptions.
	at com.google.cloud.hadoop.repackaged.gcs.com.google.cloud.hadoop.gcsio.GoogleCloudStorageExceptions.createCompositeException(GoogleCloudStorageExceptions.java:60)
	at com.google.cloud.hadoop.repackaged.gcs.com.google.cloud.hadoop.gcsio.GoogleCloudStorageImpl.getItemInfos(GoogleCloudStorageImpl.java:1612)
	at com.google.cloud.hadoop.repackaged.gcs.com.google.cloud.hadoop.gcsio.GoogleCloudStorageFileSystem.getFileInfosRaw(GoogleCloudStorageFileSystem.java:1298)
	at com.google.cloud.hadoop.repackaged.gcs.com.google.cloud.hadoop.gcsio.GoogleCloudStorageFileSystem.listFileInfo(GoogleCloudStorageFileSystem.java:1077)
	at com.google.cloud.hadoop.fs.gcs.GoogleHadoopFileSystemBase.listStatus(GoogleHadoopFileSystemBase.java:1329)
	at org.apache.hadoop.fs.generic.wrapper.GenericWrappedFileSystem.listStatus(GenericWrappedFileSystem.java:199)
	at com.hortonworks.TestSeek.run(TestSeek.java:89)
	at com.hortonworks.TestSeek.main(TestSeek.java:30)
	Suppressed: java.io.IOException: Error getting Bucket: : bucket: BUCKET_NAME
		at com.google.cloud.hadoop.repackaged.gcs.com.google.cloud.hadoop.gcsio.GoogleCloudStorageImpl.wrapException(GoogleCloudStorageImpl.java:1832)
		at com.google.cloud.hadoop.repackaged.gcs.com.google.cloud.hadoop.gcsio.GoogleCloudStorageImpl$11.onFailure(GoogleCloudStorageImpl.java:1575)
		at com.google.cloud.hadoop.repackaged.gcs.com.google.api.client.googleapis.batch.json.JsonBatchCallback.onFailure(JsonBatchCallback.java:54)
		at com.google.cloud.hadoop.repackaged.gcs.com.google.api.client.googleapis.batch.json.JsonBatchCallback.onFailure(JsonBatchCallback.java:50)
		at com.google.cloud.hadoop.repackaged.gcs.com.google.api.client.googleapis.batch.BatchUnparsedResponse.parseAndCallback(BatchUnparsedResponse.java:223)
		at com.google.cloud.hadoop.repackaged.gcs.com.google.api.client.googleapis.batch.BatchUnparsedResponse.parseNextResponse(BatchUnparsedResponse.java:155)
		at com.google.cloud.hadoop.repackaged.gcs.com.google.api.client.googleapis.batch.BatchRequest.execute(BatchRequest.java:253)
		at com.google.cloud.hadoop.repackaged.gcs.com.google.cloud.hadoop.gcsio.BatchHelper.flushIfPossible(BatchHelper.java:117)
		at com.google.cloud.hadoop.repackaged.gcs.com.google.cloud.hadoop.gcsio.BatchHelper.flush(BatchHelper.java:130)
		at com.google.cloud.hadoop.repackaged.gcs.com.google.cloud.hadoop.gcsio.GoogleCloudStorageImpl.getItemInfos(GoogleCloudStorageImpl.java:1609)
		... 6 more
	Caused by: java.io.IOException: {"code":403,"errors":[{"domain":"global","message":"SERVICE_ACCOUNT@PROJECT.iam.gserviceaccount.com does not have storage.buckets.get access to BUCKET_NAME.","reason":"forbidden"}],"message":"SERVICE_ACCOUNT@PROJECT.iam.gserviceaccount.com does not have storage.buckets.get access to BUCKET_NAME."}
		at com.google.cloud.hadoop.repackaged.gcs.com.google.cloud.hadoop.gcsio.GoogleCloudStorageImpl$11.onFailure(GoogleCloudStorageImpl.java:1573)
		... 14 more
	Suppressed: java.io.IOException: Error getting Bucket: : bucket: BUCKET_NAME
		at com.google.cloud.hadoop.repackaged.gcs.com.google.cloud.hadoop.gcsio.GoogleCloudStorageImpl.wrapException(GoogleCloudStorageImpl.java:1832)
		at com.google.cloud.hadoop.repackaged.gcs.com.google.cloud.hadoop.gcsio.GoogleCloudStorageImpl$11.onFailure(GoogleCloudStorageImpl.java:1575)
		at com.google.cloud.hadoop.repackaged.gcs.com.google.api.client.googleapis.batch.json.JsonBatchCallback.onFailure(JsonBatchCallback.java:54)
		at com.google.cloud.hadoop.repackaged.gcs.com.google.api.client.googleapis.batch.json.JsonBatchCallback.onFailure(JsonBatchCallback.java:50)
		at com.google.cloud.hadoop.repackaged.gcs.com.google.api.client.googleapis.batch.BatchUnparsedResponse.parseAndCallback(BatchUnparsedResponse.java:223)
		at com.google.cloud.hadoop.repackaged.gcs.com.google.api.client.googleapis.batch.BatchUnparsedResponse.parseNextResponse(BatchUnparsedResponse.java:155)
		at com.google.cloud.hadoop.repackaged.gcs.com.google.api.client.googleapis.batch.BatchRequest.execute(BatchRequest.java:253)
		at com.google.cloud.hadoop.repackaged.gcs.com.google.cloud.hadoop.gcsio.BatchHelper.flushIfPossible(BatchHelper.java:117)
		at com.google.cloud.hadoop.repackaged.gcs.com.google.cloud.hadoop.gcsio.BatchHelper.flush(BatchHelper.java:130)
		at com.google.cloud.hadoop.repackaged.gcs.com.google.cloud.hadoop.gcsio.GoogleCloudStorageImpl.getItemInfos(GoogleCloudStorageImpl.java:1609)
		... 6 more
	Caused by: java.io.IOException: {"code":403,"errors":[{"domain":"global","message":"SERVICE_ACCOUNT@PROJECT.iam.gserviceaccount.com does not have storage.buckets.get access to BUCKET_NAME.","reason":"forbidden"}],"message":"SERVICE_ACCOUNT@PROJECT.iam.gserviceaccount.com does not have storage.buckets.get access to BUCKET_NAME."}
		at com.google.cloud.hadoop.repackaged.gcs.com.google.cloud.hadoop.gcsio.GoogleCloudStorageImpl$11.onFailure(GoogleCloudStorageImpl.java:1573)
		... 14 more
```

--- a/bigquery/src/test/java/com/google/cloud/hadoop/io/bigquery/DynamicFileListRecordReaderTest.java
+++ b/bigquery/src/test/java/com/google/cloud/hadoop/io/bigquery/DynamicFileListRecordReaderTest.java
@@ -99,6 +99,7 @@ public class DynamicFileListRecordReaderTest {
     estimatedNumRecords = 2;
 
     fileSystem = basePath.getFileSystem(config);
+    fileSystem.mkdirs(basePath);
     fileSystem.mkdirs(shardPath.getParent());
 
     // Instead of actually blocking, make our mockSleeper throw an exception that we can catch

--- a/bigquery/src/test/java/com/google/cloud/hadoop/io/bigquery/output/FederatedBigQueryOutputCommitterTest.java
+++ b/bigquery/src/test/java/com/google/cloud/hadoop/io/bigquery/output/FederatedBigQueryOutputCommitterTest.java
@@ -84,8 +84,10 @@ public class FederatedBigQueryOutputCommitterTest {
   private static final TaskAttemptID TEST_TASK_ATTEMPT_ID =
       new TaskAttemptID(new TaskID("sample_task", 100, false, 200), 1);
 
+  private static final String TEST_BUCKET_STRING = "gs://test_bucket";
+
   /** Sample raw output path for data. */
-  private static final String TEST_OUTPUT_PATH_STRING = "gs://test_bucket/test_directory/";
+  private static final String TEST_OUTPUT_PATH_STRING = TEST_BUCKET_STRING + "/test_directory/";
 
   /** Sample output file. */
   private static final String TEST_OUTPUT_FILE_STRING = TEST_OUTPUT_PATH_STRING + "test_file";
@@ -124,6 +126,7 @@ public class FederatedBigQueryOutputCommitterTest {
 
     // Create the file system.
     ghfs = new InMemoryGoogleHadoopFileSystem();
+    ghfs.mkdirs(new Path(TEST_BUCKET_STRING));
 
     // Setup the configuration.
     job = Job.getInstance(InMemoryGoogleHadoopFileSystem.getSampleConfiguration());

--- a/bigquery/src/test/java/com/google/cloud/hadoop/io/bigquery/output/ForwardingBigQueryFileOutputCommitterTest.java
+++ b/bigquery/src/test/java/com/google/cloud/hadoop/io/bigquery/output/ForwardingBigQueryFileOutputCommitterTest.java
@@ -90,8 +90,10 @@ public class ForwardingBigQueryFileOutputCommitterTest {
   private static final TaskAttemptID TEST_TASK_ATTEMPT_ID =
       new TaskAttemptID(new TaskID("sample_task", 100, false, 200), 1);
 
+  private static final String TEST_BUCKET_STRING = "gs://test_bucket";
+
   /** Sample raw output path for data. */
-  private static final String TEST_OUTPUT_PATH_STRING = "gs://test_bucket/test_directory/";
+  private static final String TEST_OUTPUT_PATH_STRING = TEST_BUCKET_STRING + "/test_directory/";
 
   /** Sample output file. */
   private static final String TEST_OUTPUT_FILE_STRING = TEST_OUTPUT_PATH_STRING + "test_file";
@@ -125,6 +127,7 @@ public class ForwardingBigQueryFileOutputCommitterTest {
 
     // Create the file system.
     ghfs = new InMemoryGoogleHadoopFileSystem();
+    ghfs.mkdirs(new Path(TEST_BUCKET_STRING));
 
     // Setup the configuration.
     job = Job.getInstance(InMemoryGoogleHadoopFileSystem.getSampleConfiguration());

--- a/bigquery/src/test/java/com/google/cloud/hadoop/io/bigquery/output/ForwardingBigQueryFileOutputFormatTest.java
+++ b/bigquery/src/test/java/com/google/cloud/hadoop/io/bigquery/output/ForwardingBigQueryFileOutputFormatTest.java
@@ -67,8 +67,10 @@ public class ForwardingBigQueryFileOutputFormatTest {
   @SuppressWarnings("rawtypes")
   private static final Class<? extends FileOutputFormat> TEST_OUTPUT_CLASS = TextOutputFormat.class;
 
+  private static final String TEST_BUCKET_STRING = "gs://test_bucket";
+
   /** Sample raw output path for data. */
-  private static final String TEST_OUTPUT_PATH_STRING = "gs://test_bucket/test_directory/";
+  private static final String TEST_OUTPUT_PATH_STRING = TEST_BUCKET_STRING + "/test_directory/";
 
   /** Sample output path for data. */
   private static final Path TEST_OUTPUT_PATH = new Path(TEST_OUTPUT_PATH_STRING);
@@ -104,6 +106,7 @@ public class ForwardingBigQueryFileOutputFormatTest {
 
     // Create the file system.
     ghfs = new InMemoryGoogleHadoopFileSystem();
+    ghfs.mkdirs(new Path(TEST_BUCKET_STRING));
 
     // Create the configuration, but setup in the tests.
     job = Job.getInstance(InMemoryGoogleHadoopFileSystem.getSampleConfiguration());

--- a/bigquery/src/test/java/com/google/cloud/hadoop/io/bigquery/output/IndirectBigQueryOutputCommitterTest.java
+++ b/bigquery/src/test/java/com/google/cloud/hadoop/io/bigquery/output/IndirectBigQueryOutputCommitterTest.java
@@ -107,8 +107,10 @@ public class IndirectBigQueryOutputCommitterTest {
   private static final TaskAttemptID TEST_TASK_ATTEMPT_ID =
       new TaskAttemptID(new TaskID("sample_task", 100, false, 200), 1);
 
+  private static final String TEST_BUCKET_STRING = "gs://test_bucket";
+
   /** Sample raw output path for data. */
-  private static final String TEST_OUTPUT_PATH_STRING = "gs://test_bucket/test_directory/";
+  private static final String TEST_OUTPUT_PATH_STRING = TEST_BUCKET_STRING + "/test_directory/";
 
   /** Sample output file. */
   private static final String TEST_OUTPUT_FILE_STRING = TEST_OUTPUT_PATH_STRING + "test_file";
@@ -147,6 +149,7 @@ public class IndirectBigQueryOutputCommitterTest {
 
     // Create the file system.
     ghfs = new InMemoryGoogleHadoopFileSystem();
+    ghfs.mkdirs(new Path(TEST_BUCKET_STRING));
 
     // Setup the configuration.
     job = Job.getInstance(InMemoryGoogleHadoopFileSystem.getSampleConfiguration());

--- a/gcs/CHANGES.md
+++ b/gcs/CHANGES.md
@@ -132,8 +132,9 @@
     fs.gs.glob.algorithm (default: CONCURRENT)
     ```
 
-1.  Do not create the parent directory objects when writing a file, instead rely
-    on the implicit directory inference.
+1.  Do not create the parent directory objects (this includes buckets) when
+    creating a new file or a directory, instead rely on the implicit directory
+    inference.
 
 1.  Use default logging backend for Google Flogger instead of Slf4j.
 

--- a/gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/HadoopFileSystemTestBase.java
+++ b/gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/HadoopFileSystemTestBase.java
@@ -73,6 +73,8 @@ public abstract class HadoopFileSystemTestBase extends GoogleCloudStorageFileSys
   public static void postCreateInit(HadoopFileSystemIntegrationHelper helper)
       throws IOException {
     ghfsHelper = helper;
+    ghfsHelper.ghfs.mkdirs(
+        new Path(ghfsHelper.ghfsFileSystemDescriptor.getFileSystemRoot().toUri()));
     GoogleCloudStorageFileSystemIntegrationTest.postCreateInit(ghfsHelper);
 
     // Ensures that we do not accidentally end up testing wrong functionality.
@@ -133,16 +135,19 @@ public abstract class HadoopFileSystemTestBase extends GoogleCloudStorageFileSys
 
       Instant currentTime = Instant.now();
       Instant modificationTime = Instant.ofEpochMilli(fileStatus.getModificationTime());
-      // We must subtract 1000, because some FileSystems, like LocalFileSystem, have only
-      // second granularity, so we might have something like testStartTime == 1234123
-      // and modificationTime == 1234000. Unfortunately, "Instant" doesn't support easy
-      // conversions between units to clip to the "second" precision.
-      // Alternatively, we should just use TimeUnit and formally convert "toSeconds".
-      assertWithMessage(
-              "Stale file? testStartTime: %s modificationTime: %s bucket: '%s' object: '%s'",
-              testStartTime, modificationTime, bucketName, objectName)
-          .that(modificationTime)
-          .isAtLeast(testStartTime.minusMillis(1000));
+      // Ignore modification time for inferred directories that always set to 0
+      if (!expectedToBeDir || fileStatus.getModificationTime() != 0) {
+        // We must subtract 1000, because some FileSystems, like LocalFileSystem, have only
+        // second granularity, so we might have something like testStartTime == 1234123
+        // and modificationTime == 1234000. Unfortunately, "Instant" doesn't support easy
+        // conversions between units to clip to the "second" precision.
+        // Alternatively, we should just use TimeUnit and formally convert "toSeconds".
+        assertWithMessage(
+                "Stale file? testStartTime: %s modificationTime: %s bucket: '%s' object: '%s'",
+                testStartTime, modificationTime, bucketName, objectName)
+            .that(modificationTime)
+            .isAtLeast(testStartTime.minusMillis(1000));
+      }
       assertWithMessage(
               "Clock skew? currentTime: %s modificationTime: %s bucket: '%s' object: '%s'",
               currentTime, modificationTime, bucketName, objectName)

--- a/gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/WebHdfsIntegrationTest.java
+++ b/gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/WebHdfsIntegrationTest.java
@@ -145,16 +145,19 @@ public class WebHdfsIntegrationTest extends HadoopFileSystemTestBase {
   }
 
   /**
-   * Validates that we cannot open a non-existent object.
-   * Note: WebHDFS throws IOException (Internal Server Error (error code=500))
+   * Validates that we cannot open a non-existent object. Note: WebHDFS throws IOException (Internal
+   * Server Error (error code=500))
    */
-  @Test @Override
-  public void testOpenNonExistent()
-      throws IOException {
-    String bucketName = ghfsHelper.getUniqueBucketName("open-non-existent");
+  @Test
+  @Override
+  public void testOpenNonExistentObject() throws IOException {
+    String bucketName = ghfsHelper.sharedBucketName1;
     IOException e =
         assertThrows(
-            IOException.class, () -> ghfsHelper.readTextFile(bucketName, objectName, 0, 100, true));
+            IOException.class,
+            () ->
+                ghfsHelper.readTextFile(
+                    bucketName, objectName + "_open-non-existent", 0, 100, true));
     assertThat(e).hasMessageThat().contains("Internal Server Error (error code=500)");
   }
 

--- a/gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/WebHdfsIntegrationTest.java
+++ b/gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/WebHdfsIntegrationTest.java
@@ -170,7 +170,7 @@ public class WebHdfsIntegrationTest extends HadoopFileSystemTestBase {
   @Test
   @Override
   public void testOpenInNonExistentBucket() throws IOException {
-    String bucketName = ghfsHelper.sharedBucketName1 + "_open-non-existent";
+    String bucketName = ghfsHelper.getUniqueBucketName("open-non-existent");
     IOException e =
         assertThrows(
             IOException.class, () -> ghfsHelper.readTextFile(bucketName, objectName, 0, 100, true));

--- a/gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/WebHdfsIntegrationTest.java
+++ b/gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/WebHdfsIntegrationTest.java
@@ -145,8 +145,9 @@ public class WebHdfsIntegrationTest extends HadoopFileSystemTestBase {
   }
 
   /**
-   * Validates that we cannot open a non-existent object. Note: WebHDFS throws IOException (Internal
-   * Server Error (error code=500))
+   * Validates that we cannot open a non-existent object.
+   *
+   * <p>Note: WebHDFS throws IOException (Internal Server Error (error code=500))
    */
   @Test
   @Override
@@ -158,6 +159,21 @@ public class WebHdfsIntegrationTest extends HadoopFileSystemTestBase {
             () ->
                 ghfsHelper.readTextFile(
                     bucketName, objectName + "_open-non-existent", 0, 100, true));
+    assertThat(e).hasMessageThat().contains("Internal Server Error (error code=500)");
+  }
+
+  /**
+   * Validates that we cannot open an object in non-existent bucket.
+   *
+   * <p>Note: WebHDFS throws IOException (Internal Server Error (error code=500))
+   */
+  @Test
+  @Override
+  public void testOpenInNonExistentBucket() throws IOException {
+    String bucketName = ghfsHelper.sharedBucketName1 + "_open-non-existent";
+    IOException e =
+        assertThrows(
+            IOException.class, () -> ghfsHelper.readTextFile(bucketName, objectName, 0, 100, true));
     assertThat(e).hasMessageThat().contains("Internal Server Error (error code=500)");
   }
 

--- a/gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/contract/GoogleContract.java
+++ b/gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/contract/GoogleContract.java
@@ -22,7 +22,10 @@ import static com.google.cloud.hadoop.util.HadoopCredentialConfiguration.SERVICE
 
 import com.google.cloud.hadoop.gcsio.integration.GoogleCloudStorageTestHelper.TestBucketHelper;
 import com.google.cloud.hadoop.gcsio.testing.TestConfiguration;
+import java.io.IOException;
 import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
 import org.apache.hadoop.fs.contract.AbstractBondedFSContract;
 
 /** Contract of GoogleHadoopFileSystem via scheme "gs". */
@@ -49,6 +52,13 @@ public class GoogleContract extends AbstractBondedFSContract {
           GCS_CONFIG_PREFIX + SERVICE_ACCOUNT_KEYFILE_SUFFIX.getKey(),
           testConf.getPrivateKeyFile());
     }
+  }
+
+  @Override
+  public void init() throws IOException {
+    super.init();
+    FileSystem testFs = getTestFileSystem();
+    testFs.mkdirs(new Path(testFs.getUri()));
   }
 
   @Override

--- a/gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageFileSystem.java
+++ b/gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageFileSystem.java
@@ -447,7 +447,8 @@ public class GoogleCloudStorageFileSystem {
       try {
         gcs.createBucket(resourceId.getBucketName());
       } catch (IOException e) {
-        if (ApiErrorExtractor.INSTANCE.itemAlreadyExists(e)) {
+        if (e instanceof FileAlreadyExistsException
+            || ApiErrorExtractor.INSTANCE.itemAlreadyExists(e)) {
           // This means that bucket already exist and we do not need to do anything.
           logger.atFine().withCause(e).log(
               "mkdirs: %s already exists, ignoring creation failure", resourceId);

--- a/gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageFileSystem.java
+++ b/gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageFileSystem.java
@@ -18,8 +18,10 @@ package com.google.cloud.hadoop.gcsio;
 
 import static com.google.cloud.hadoop.gcsio.GoogleCloudStorage.PATH_DELIMITER;
 import static com.google.common.base.Preconditions.checkArgument;
-import static com.google.common.base.Preconditions.checkNotNull;
 import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.base.Strings.isNullOrEmpty;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static com.google.common.util.concurrent.Futures.immediateFuture;
 import static com.google.common.util.concurrent.MoreExecutors.newDirectExecutorService;
 import static java.lang.Math.min;
 import static java.util.Comparator.comparing;
@@ -29,10 +31,10 @@ import com.google.api.client.auth.oauth2.Credential;
 import com.google.cloud.hadoop.gcsio.GoogleCloudStorage.ListPage;
 import com.google.cloud.hadoop.gcsio.cooplock.CoopLockOperationDelete;
 import com.google.cloud.hadoop.gcsio.cooplock.CoopLockOperationRename;
+import com.google.cloud.hadoop.util.ApiErrorExtractor;
 import com.google.cloud.hadoop.util.LazyExecutorService;
 import com.google.common.annotations.VisibleForTesting;
 import com.google.common.base.Preconditions;
-import com.google.common.base.Strings;
 import com.google.common.collect.ImmutableList;
 import com.google.common.collect.Iterables;
 import com.google.common.collect.Lists;
@@ -219,8 +221,8 @@ public class GoogleCloudStorageFileSystem {
    * @param path Object full path of the form gs://bucket/object-path.
    * @return A channel for writing to the given object.
    */
-  public WritableByteChannel create(URI path, CreateFileOptions options) throws IOException {
-    logger.atFine().log("create(path: %s, options: %s)", path, options);
+  public WritableByteChannel create(URI path, CreateFileOptions createOptions) throws IOException {
+    logger.atFine().log("create(path: %s, createOptions: %s)", path, createOptions);
     Preconditions.checkNotNull(path, "path could not be null");
     StorageResourceId resourceId =
         StorageResourceId.fromUriPath(path, /* allowEmptyObjectName=*/ true);
@@ -231,21 +233,34 @@ public class GoogleCloudStorageFileSystem {
               "Cannot create a file whose name looks like a directory: '%s'", resourceId));
     }
 
-    // Check if a directory of that name exists.
-    if (options.isEnsureNoDirectoryConflict()
-        && getFileInfoInternal(resourceId.toDirectoryId(), /* inferImplicitDirectories */ true)
-            .exists()) {
-      throw new FileAlreadyExistsException("A directory with that name exists: " + path);
+    // Before creating a top-level directory we need to check if there are no conflicting files
+    // with the same name as any subdirectory
+    if (options.isEnsureNoConflictingItems()) {
+      // Asynchronously check if a directory with the same name exists.
+      StorageResourceId dirId = resourceId.toDirectoryId();
+      Future<Boolean> conflictingDirExist =
+          createOptions.isEnsureNoDirectoryConflict()
+              ? cachedExecutor.submit(
+                  () -> getFileInfoInternal(dirId, /* inferImplicitDirectories */ true).exists())
+              : immediateFuture(false);
+
+      checkNoFilesConflictingWithDirs(resourceId);
+
+      // Check if a directory of that name exists.
+      if (getFromFuture(conflictingDirExist)) {
+        throw new FileAlreadyExistsException("A directory with that name exists: " + path);
+      }
     }
 
-    if (options.getOverwriteGenerationId() != StorageResourceId.UNKNOWN_GENERATION_ID) {
+    if (createOptions.getOverwriteGenerationId() != StorageResourceId.UNKNOWN_GENERATION_ID) {
       resourceId =
           new StorageResourceId(
               resourceId.getBucketName(),
               resourceId.getObjectName(),
-              options.getOverwriteGenerationId());
+              createOptions.getOverwriteGenerationId());
     }
-    return gcs.create(resourceId, objectOptionsFromFileOptions(options));
+
+    return gcs.create(resourceId, objectOptionsFromFileOptions(createOptions));
   }
 
   /**
@@ -421,64 +436,45 @@ public class GoogleCloudStorageFileSystem {
     mkdirsInternal(StorageResourceId.fromUriPath(path, /* allowEmptyObjectName= */ true));
   }
 
-  public void mkdirsInternal(StorageResourceId resourceId) throws IOException {
+  private void mkdirsInternal(StorageResourceId resourceId) throws IOException {
     if (resourceId.isRoot()) {
       // GCS_ROOT directory always exists, no need to go through the rest of the method.
       return;
     }
 
+    // In case path is a bucket we just attempt to create it without additional checks
+    if (resourceId.isBucket()) {
+      try {
+        gcs.createBucket(resourceId.getBucketName());
+      } catch (FileAlreadyExistsException e) {
+        // This means that bucket already exist and we do not need to do anything.
+        logger.atFine().withCause(e).log(
+            "mkdirs: %s already exists, ignoring creation failure", resourceId);
+      }
+      return;
+    }
+
     resourceId = resourceId.toDirectoryId();
 
-    // Create a list of all intermediate paths.
-    // For example,
-    // gs://foo/bar/zoo/ => (gs://foo/, gs://foo/bar/, gs://foo/bar/zoo/)
-    //
-    // We also need to find out if any of the subdir path item exists as a file
-    // therefore, also create a list of file equivalent of each subdir path.
-    // For example,
-    // gs://foo/bar/zoo/ => (gs://foo/bar, gs://foo/bar/zoo)
-    List<String> subdirs = getSubDirs(resourceId.getObjectName());
-    List<StorageResourceId> itemIds = new ArrayList<>(subdirs.size() * 2 + 1);
-    for (String subdir : subdirs) {
-      itemIds.add(new StorageResourceId(resourceId.getBucketName(), subdir));
-      if (!Strings.isNullOrEmpty(subdir)) {
-        itemIds.add(
-            new StorageResourceId(resourceId.getBucketName(), StringPaths.toFilePath(subdir)));
-      }
-    }
-    // Add the bucket portion.
-    itemIds.add(new StorageResourceId(resourceId.getBucketName()));
-    logger.atFiner().log("mkdirs: going to create dirs with %s paths", itemIds);
-
-    List<GoogleCloudStorageItemInfo> itemInfos = gcs.getItemInfos(itemIds);
-
-    // Each intermediate path must satisfy one of the following conditions:
-    // -- it does not exist or
-    // -- if it exists, it is a directory
-    //
-    // If any of the intermediate paths violates these requirements then
-    // bail out early so that we do not end up with partial set of
-    // created sub-directories. It is possible that the status of intermediate
-    // paths can change after we make this check therefore this is a
-    // good faith effort and not a guarantee.
-    GoogleCloudStorageItemInfo bucketInfo = null;
-    List<StorageResourceId> subdirsToCreate = new ArrayList<>(subdirs.size());
-    for (GoogleCloudStorageItemInfo info : itemInfos) {
-      if (info.isBucket()) {
-        checkState(bucketInfo == null, "bucketInfo should be null");
-        bucketInfo = info;
-      } else if (info.getResourceId().isDirectory() && !info.exists()) {
-        subdirsToCreate.add(info.getResourceId());
-      } else if (!info.getResourceId().isDirectory() && info.exists()) {
-        throw new FileAlreadyExistsException(
-            "Cannot create directories because of existing file: " + info.getResourceId());
-      }
+    // Before creating a top-level directory we need to check if there are no conflicting files
+    // with the same name as any subdirectory
+    if (options.isEnsureNoConflictingItems()) {
+      checkNoFilesConflictingWithDirs(resourceId);
     }
 
-    if (!checkNotNull(bucketInfo, "bucketInfo should not be null").exists()) {
-      gcs.createBucket(bucketInfo.getBucketName());
+    // Create only a top-level directory because subdirectories will be inferred
+    // if top-level directory exists
+    try {
+      gcs.createEmptyObject(resourceId);
+    } catch (FileAlreadyExistsException e) {
+      // This means that directory object already exist and we do not need to do anything.
+      logger.atFine().withCause(e).log(
+          "mkdirs: %s already exists, ignoring creation failure", resourceId);
+    } catch (IOException e) {
+      throw ApiErrorExtractor.INSTANCE.itemNotFound(e)
+          ? new IOException(String.format("Failed to create: '%s'", resourceId), e)
+          : e;
     }
-    gcs.createEmptyObjects(subdirsToCreate);
   }
 
   /**
@@ -936,14 +932,14 @@ public class GoogleCloudStorageFileSystem {
     }
   }
 
-  private static <T> T getFromFuture(Future<T> future) throws IOException {
+  static <T> T getFromFuture(Future<T> future) throws IOException {
     try {
       return future.get();
-    } catch (InterruptedException | ExecutionException e) {
+    } catch (ExecutionException | InterruptedException e) {
       if (e instanceof InterruptedException) {
         Thread.currentThread().interrupt();
       }
-      throw new IOException("Failed to get info from future", e);
+      throw new IOException("Failed to get result", e);
     }
   }
 
@@ -1101,15 +1097,7 @@ public class GoogleCloudStorageFileSystem {
       }
     }
 
-    List<GoogleCloudStorageItemInfo> listDirInfo;
-    try {
-      listDirInfo = listDirFuture.get();
-    } catch (ExecutionException | InterruptedException e) {
-      if (e instanceof InterruptedException) {
-        Thread.currentThread().interrupt();
-      }
-      throw new IOException(String.format("Failed to get dir info for '%s'", resourceId), e);
-    }
+    List<GoogleCloudStorageItemInfo> listDirInfo = getFromFuture(listDirFuture);
     if (listDirInfo.isEmpty()) {
       return GoogleCloudStorageItemInfo.createNotFound(resourceId);
     }
@@ -1154,15 +1142,7 @@ public class GoogleCloudStorageFileSystem {
 
       List<FileInfo> infos = new ArrayList<>(paths.size());
       for (Future<FileInfo> infoFuture : infoFutures) {
-        try {
-          infos.add(infoFuture.get());
-        } catch (InterruptedException | ExecutionException e) {
-          if (e instanceof InterruptedException) {
-            Thread.currentThread().interrupt();
-          }
-          throw new IOException(
-              String.format("Failed to getFileInfos for %d paths", paths.size()), e);
-        }
+        infos.add(getFromFuture(infoFuture));
       }
       return infos;
     } finally {
@@ -1222,38 +1202,61 @@ public class GoogleCloudStorageFileSystem {
     gcs.createEmptyObject(resourceId);
   }
 
+  private void checkNoFilesConflictingWithDirs(StorageResourceId resourceId) throws IOException {
+    // Create a list of all files that can conflict with intermediate/subdirectory paths.
+    // For example: gs://foo/bar/zoo/ => (gs://foo/bar, gs://foo/bar/zoo)
+    List<StorageResourceId> fileIds =
+        getDirs(resourceId.getObjectName()).stream()
+            .filter(subdir -> !isNullOrEmpty(subdir))
+            .map(
+                subdir ->
+                    new StorageResourceId(
+                        resourceId.getBucketName(), StringPaths.toFilePath(subdir)))
+            .collect(toImmutableList());
+
+    // Each intermediate path must ensure that corresponding file does not exist
+    //
+    // If for any of the intermediate paths file already exists then bail out early.
+    // It is possible that the status of intermediate paths can change after
+    // we make this check therefore this is a good faith effort and not a guarantee.
+    for (GoogleCloudStorageItemInfo fileInfo : gcs.getItemInfos(fileIds)) {
+      if (fileInfo.exists()) {
+        throw new FileAlreadyExistsException(
+            "Cannot create directories because of existing file: " + fileInfo.getResourceId());
+      }
+    }
+  }
+
   /**
-   * For objects whose name looks like a path (foo/bar/zoo), returns intermediate sub-paths.
+   * For objects whose name looks like a path (foo/bar/zoo), returns all directory paths.
    *
    * <p>For example:
    *
    * <ul>
    *   <li>foo/bar/zoo => returns: (foo/, foo/bar/)
+   *   <li>foo/bar/zoo/ => returns: (foo/, foo/bar/, foo/bar/zoo/)
    *   <li>foo => returns: ()
    * </ul>
    *
    * @param objectName Name of an object.
    * @return List of sub-directory like paths.
    */
-  static List<String> getSubDirs(String objectName) {
-    List<String> subdirs = new ArrayList<>();
-    if (!Strings.isNullOrEmpty(objectName)) {
-      int currentIndex = 0;
-      while (currentIndex < objectName.length()) {
-        int index = objectName.indexOf(PATH_DELIMITER, currentIndex);
-        if (index < 0) {
-          break;
-        }
-        subdirs.add(objectName.substring(0, index + PATH_DELIMITER.length()));
-        currentIndex = index + PATH_DELIMITER.length();
-      }
+  static List<String> getDirs(String objectName) {
+    if (isNullOrEmpty(objectName)) {
+      return ImmutableList.of();
     }
-    return subdirs;
+    List<String> dirs = new ArrayList<>();
+    int index = 0;
+    while ((index = objectName.indexOf(PATH_DELIMITER, index)) >= 0) {
+      index = index + PATH_DELIMITER.length();
+      dirs.add(objectName.substring(0, index));
+    }
+    return dirs;
   }
 
   /** Gets the leaf item of the given path. */
   String getItemName(URI path) {
-    Preconditions.checkNotNull(path);
+    Preconditions.checkNotNull(path, "path can not be null");
 
     // There is no leaf item for the root path.
     if (path.equals(GCS_ROOT)) {

--- a/gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageFileSystem.java
+++ b/gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageFileSystem.java
@@ -233,7 +233,7 @@ public class GoogleCloudStorageFileSystem {
               "Cannot create a file whose name looks like a directory: '%s'", resourceId));
     }
 
-    // Before creating a top-level directory we need to check if there are no conflicting files
+    // Before creating a leaf directory we need to check if there are no conflicting files
     // with the same name as any subdirectory
     if (options.isEnsureNoConflictingItems()) {
       // Asynchronously check if a directory with the same name exists.
@@ -446,29 +446,24 @@ public class GoogleCloudStorageFileSystem {
     if (resourceId.isBucket()) {
       try {
         gcs.createBucket(resourceId.getBucketName());
-      } catch (IOException e) {
-        if (e instanceof FileAlreadyExistsException
-            || ApiErrorExtractor.INSTANCE.itemAlreadyExists(e)) {
-          // This means that bucket already exist and we do not need to do anything.
-          logger.atFine().withCause(e).log(
-              "mkdirs: %s already exists, ignoring creation failure", resourceId);
-        } else {
-          throw e;
-        }
+      } catch (FileAlreadyExistsException e) {
+        // This means that bucket already exist and we do not need to do anything.
+        logger.atFine().withCause(e).log(
+            "mkdirs: %s already exists, ignoring creation failure", resourceId);
       }
       return;
     }
 
     resourceId = resourceId.toDirectoryId();
 
-    // Before creating a top-level directory we need to check if there are no conflicting files
+    // Before creating a leaf directory we need to check if there are no conflicting files
     // with the same name as any subdirectory
     if (options.isEnsureNoConflictingItems()) {
       checkNoFilesConflictingWithDirs(resourceId);
     }
 
-    // Create only a top-level directory because subdirectories will be inferred
-    // if top-level directory exists
+    // Create only a leaf directory because subdirectories will be inferred
+    // if leaf directory exists
     try {
       gcs.createEmptyObject(resourceId);
     } catch (IOException e) {

--- a/gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageFileSystem.java
+++ b/gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageFileSystem.java
@@ -446,10 +446,14 @@ public class GoogleCloudStorageFileSystem {
     if (resourceId.isBucket()) {
       try {
         gcs.createBucket(resourceId.getBucketName());
-      } catch (FileAlreadyExistsException e) {
-        // This means that bucket already exist and we do not need to do anything.
-        logger.atFine().withCause(e).log(
-            "mkdirs: %s already exists, ignoring creation failure", resourceId);
+      } catch (IOException e) {
+        if (ApiErrorExtractor.INSTANCE.itemAlreadyExists(e)) {
+          // This means that bucket already exist and we do not need to do anything.
+          logger.atFine().withCause(e).log(
+              "mkdirs: %s already exists, ignoring creation failure", resourceId);
+        } else {
+          throw e;
+        }
       }
       return;
     }
@@ -466,14 +470,13 @@ public class GoogleCloudStorageFileSystem {
     // if top-level directory exists
     try {
       gcs.createEmptyObject(resourceId);
-    } catch (FileAlreadyExistsException e) {
-      // This means that directory object already exist and we do not need to do anything.
-      logger.atFine().withCause(e).log(
-          "mkdirs: %s already exists, ignoring creation failure", resourceId);
     } catch (IOException e) {
-      throw ApiErrorExtractor.INSTANCE.itemNotFound(e)
-          ? new IOException(String.format("Failed to create: '%s'", resourceId), e)
-          : e;
+      if (ApiErrorExtractor.INSTANCE.itemAlreadyExists(e)) {
+        logger.atFine().withCause(e).log(
+            "mkdirs: %s already exists, ignoring creation failure", resourceId);
+      } else {
+        throw e;
+      }
     }
   }
 

--- a/gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageFileSystem.java
+++ b/gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageFileSystem.java
@@ -31,7 +31,6 @@ import com.google.api.client.auth.oauth2.Credential;
 import com.google.cloud.hadoop.gcsio.GoogleCloudStorage.ListPage;
 import com.google.cloud.hadoop.gcsio.cooplock.CoopLockOperationDelete;
 import com.google.cloud.hadoop.gcsio.cooplock.CoopLockOperationRename;
-import com.google.cloud.hadoop.util.ApiErrorExtractor;
 import com.google.cloud.hadoop.util.LazyExecutorService;
 import com.google.common.annotations.VisibleForTesting;
 import com.google.common.base.Preconditions;
@@ -466,13 +465,10 @@ public class GoogleCloudStorageFileSystem {
     // if leaf directory exists
     try {
       gcs.createEmptyObject(resourceId);
-    } catch (IOException e) {
-      if (ApiErrorExtractor.INSTANCE.itemAlreadyExists(e)) {
-        logger.atFine().withCause(e).log(
-            "mkdirs: %s already exists, ignoring creation failure", resourceId);
-      } else {
-        throw e;
-      }
+    } catch (FileAlreadyExistsException e) {
+      // This means that directory object already exist and we do not need to do anything.
+      logger.atFine().withCause(e).log(
+          "mkdirs: %s already exists, ignoring creation failure", resourceId);
     }
   }
 

--- a/gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageFileSystemOptions.java
+++ b/gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageFileSystemOptions.java
@@ -29,7 +29,7 @@ public abstract class GoogleCloudStorageFileSystemOptions {
         .setCloudStorageOptions(GoogleCloudStorageOptions.DEFAULT)
         .setBucketDeleteEnabled(false)
         .setMarkerFilePattern((String) null)
-        .setStatusParallelEnabled(false)
+        .setStatusParallelEnabled(true)
         .setCooperativeLockingEnabled(false);
   }
 
@@ -49,6 +49,8 @@ public abstract class GoogleCloudStorageFileSystemOptions {
   public abstract boolean isStatusParallelEnabled();
 
   public abstract boolean isCooperativeLockingEnabled();
+
+  public abstract boolean isEnsureNoConflictingItems();
 
   public void throwIfNotValid() {
     getCloudStorageOptions().throwIfNotValid();

--- a/gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageFileSystemOptions.java
+++ b/gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageFileSystemOptions.java
@@ -30,7 +30,8 @@ public abstract class GoogleCloudStorageFileSystemOptions {
         .setBucketDeleteEnabled(false)
         .setMarkerFilePattern((String) null)
         .setStatusParallelEnabled(true)
-        .setCooperativeLockingEnabled(false);
+        .setCooperativeLockingEnabled(false)
+        .setEnsureNoConflictingItems(true);
   }
 
   public abstract Builder toBuilder();
@@ -83,6 +84,8 @@ public abstract class GoogleCloudStorageFileSystemOptions {
     public abstract Builder setStatusParallelEnabled(boolean statusParallelEnabled);
 
     public abstract Builder setCooperativeLockingEnabled(boolean cooperativeLockingEnabled);
+
+    public abstract Builder setEnsureNoConflictingItems(boolean ensureNoConflictingItems);
 
     public abstract GoogleCloudStorageFileSystemOptions build();
   }

--- a/gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageImpl.java
+++ b/gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageImpl.java
@@ -464,6 +464,13 @@ public class GoogleCloudStorageImpl implements GoogleCloudStorage {
     } catch (InterruptedException e) {
       Thread.currentThread().interrupt();
       throw new IOException("Failed to create bucket", e);
+    } catch (IOException e) {
+      if (ApiErrorExtractor.INSTANCE.itemAlreadyExists(e)) {
+        throw (FileAlreadyExistsException)
+            new FileAlreadyExistsException(String.format("Bucket '%s' already exists.", bucketName))
+                .initCause(e);
+      }
+      throw e;
     }
   }
 
@@ -476,12 +483,18 @@ public class GoogleCloudStorageImpl implements GoogleCloudStorage {
     Storage.Objects.Insert insertObject = prepareEmptyInsert(resourceId, options);
     try {
       insertObject.execute();
-    } catch (IOException ioe) {
-      if (canIgnoreExceptionForEmptyObject(ioe, resourceId, options)) {
-        logger.atInfo().withCause(ioe).log(
+    } catch (IOException e) {
+      if (canIgnoreExceptionForEmptyObject(e, resourceId, options)) {
+        logger.atInfo().withCause(e).log(
             "Ignoring exception; verified object already exists with desired state.");
       } else {
-        throw ioe;
+        if (ApiErrorExtractor.INSTANCE.itemAlreadyExists(e)) {
+          throw (FileAlreadyExistsException)
+              new FileAlreadyExistsException(
+                      String.format("Object '%s' already exists.", resourceId))
+                  .initCause(e);
+        }
+        throw e;
       }
     }
   }
@@ -1968,19 +1981,17 @@ public class GoogleCloudStorageImpl implements GoogleCloudStorage {
       // maximum delay that caller will wait to receive an exception in the case of an incorrect
       // assumption and this being a scenario other than the multiple workers racing situation.
       GoogleCloudStorageItemInfo existingInfo;
-      BackOff backOff;
       int maxWaitMillis = storageOptions.getMaxWaitMillisForEmptyObjectCreation();
-      if (maxWaitMillis > 0) {
-        backOff = new ExponentialBackOff.Builder()
-            .setMaxElapsedTimeMillis(maxWaitMillis)
-            .setMaxIntervalMillis(500)
-            .setInitialIntervalMillis(100)
-            .setMultiplier(1.5)
-            .setRandomizationFactor(0.15)
-            .build();
-      } else {
-        backOff = BackOff.STOP_BACKOFF;
-      }
+      BackOff backOff =
+          maxWaitMillis > 0
+              ? new ExponentialBackOff.Builder()
+                  .setMaxElapsedTimeMillis(maxWaitMillis)
+                  .setMaxIntervalMillis(500)
+                  .setInitialIntervalMillis(100)
+                  .setMultiplier(1.5)
+                  .setRandomizationFactor(0.15)
+                  .build()
+              : BackOff.STOP_BACKOFF;
       long nextSleep = 0L;
       do {
         if (nextSleep > 0) {

--- a/gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageImpl.java
+++ b/gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageImpl.java
@@ -24,6 +24,7 @@ import static com.google.common.base.Preconditions.checkNotNull;
 import static com.google.common.base.Preconditions.checkState;
 import static com.google.common.base.Strings.isNullOrEmpty;
 import static com.google.common.collect.Sets.newConcurrentHashSet;
+import static java.lang.Math.toIntExact;
 
 import com.google.api.client.auth.oauth2.Credential;
 import com.google.api.client.googleapis.batch.json.JsonBatchCallback;
@@ -442,15 +443,11 @@ public class GoogleCloudStorageImpl implements GoogleCloudStorage {
             .setLocation(options.getLocation())
             .setStorageClass(options.getStorageClass());
     if (options.getTtl() != null) {
-      bucket.setLifecycle(
-          new Lifecycle()
-              .setRule(
-                  ImmutableList.of(
-                      new Rule()
-                          .setAction(new Action().setType("Delete"))
-                          .setCondition(
-                              new Condition()
-                                  .setAge(Math.toIntExact(options.getTtl().toDays()))))));
+      Rule lifecycleRule =
+          new Rule()
+              .setAction(new Action().setType("Delete"))
+              .setCondition(new Condition().setAge(toIntExact(options.getTtl().toDays())));
+      bucket.setLifecycle(new Lifecycle().setRule(ImmutableList.of(lifecycleRule)));
     }
 
     Storage.Buckets.Insert insertBucket =

--- a/gcsio/src/main/java/com/google/cloud/hadoop/gcsio/testing/InMemoryGoogleCloudStorage.java
+++ b/gcsio/src/main/java/com/google/cloud/hadoop/gcsio/testing/InMemoryGoogleCloudStorage.java
@@ -155,14 +155,13 @@ public class InMemoryGoogleCloudStorage implements GoogleCloudStorage {
     if (!validateBucketName(bucketName)) {
       throw new IOException("Error creating bucket. Invalid name: " + bucketName);
     }
-    if (!bucketLookup.containsKey(bucketName)) {
-      bucketLookup.put(
-          bucketName,
-          new InMemoryBucketEntry(
-              bucketName, clock.currentTimeMillis(), clock.currentTimeMillis(), options));
-    } else {
-      throw new IOException("Bucket '" + bucketName + "'already exists");
+    if (bucketLookup.containsKey(bucketName)) {
+      throw new FileAlreadyExistsException("Bucket '" + bucketName + "' already exists");
     }
+    bucketLookup.put(
+        bucketName,
+        new InMemoryBucketEntry(
+            bucketName, clock.currentTimeMillis(), clock.currentTimeMillis(), options));
   }
 
   @Override

--- a/gcsio/src/test/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageFileSystemIntegrationTest.java
+++ b/gcsio/src/test/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageFileSystemIntegrationTest.java
@@ -1622,7 +1622,7 @@ public class GoogleCloudStorageFileSystemIntegrationTest {
   private List<URI> getSubDirPaths(URI path) {
     StorageResourceId resourceId = StorageResourceId.fromUriPath(path, true);
 
-    List<String> subdirs = GoogleCloudStorageFileSystem.getSubDirs(resourceId.getObjectName());
+    List<String> subdirs = GoogleCloudStorageFileSystem.getDirs(resourceId.getObjectName());
     List<URI> subDirPaths = new ArrayList<>(subdirs.size());
     for (String subdir : subdirs) {
       subDirPaths.add(gcsiHelper.getPath(resourceId.getBucketName(), subdir));

--- a/gcsio/src/test/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageFileSystemIntegrationTest.java
+++ b/gcsio/src/test/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageFileSystemIntegrationTest.java
@@ -596,6 +596,15 @@ public class GoogleCloudStorageFileSystemIntegrationTest {
         () -> gcsiHelper.readTextFile(bucketName, objectName + "_open-non-existent", 0, 100, true));
   }
 
+  /** Validates that we cannot open an object in non-existent bucket. */
+  @Test
+  public void testOpenInNonExistentBucket() throws IOException {
+    String bucketName = sharedBucketName1 + "_open-non-existent";
+    assertThrows(
+        FileNotFoundException.class,
+        () -> gcsiHelper.readTextFile(bucketName, objectName, 0, 100, true));
+  }
+
   /** Validates delete(). */
   public void deleteHelper(DeletionBehavior behavior) throws Exception {
     String bucketName = sharedBucketName1;

--- a/gcsio/src/test/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageFileSystemIntegrationTest.java
+++ b/gcsio/src/test/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageFileSystemIntegrationTest.java
@@ -599,7 +599,7 @@ public class GoogleCloudStorageFileSystemIntegrationTest {
   /** Validates that we cannot open an object in non-existent bucket. */
   @Test
   public void testOpenInNonExistentBucket() throws IOException {
-    String bucketName = sharedBucketName1 + "_open-non-existent";
+    String bucketName = gcsiHelper.getUniqueBucketName("open-non-existent");
     assertThrows(
         FileNotFoundException.class,
         () -> gcsiHelper.readTextFile(bucketName, objectName, 0, 100, true));

--- a/gcsio/src/test/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageFileSystemIntegrationTest.java
+++ b/gcsio/src/test/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageFileSystemIntegrationTest.java
@@ -587,16 +587,13 @@ public class GoogleCloudStorageFileSystemIntegrationTest {
     }
   }
 
-  /**
-   * Validates that we cannot open a non-existent object.
-   */
+  /** Validates that we cannot open a non-existent object. */
   @Test
-  public void testOpenNonExistent()
-      throws IOException {
-    String bucketName = gcsiHelper.getUniqueBucketName("open-non-existent");
+  public void testOpenNonExistentObject() throws IOException {
+    String bucketName = sharedBucketName1;
     assertThrows(
         FileNotFoundException.class,
-        () -> gcsiHelper.readTextFile(bucketName, objectName, 0, 100, true));
+        () -> gcsiHelper.readTextFile(bucketName, objectName + "_open-non-existent", 0, 100, true));
   }
 
   /** Validates delete(). */
@@ -832,8 +829,12 @@ public class GoogleCloudStorageFileSystemIntegrationTest {
 
     // Make a path where the bucket is a non-existent parent directory.
     String uniqueBucketName2 = gcsiHelper.getUniqueBucketName("mkdir-2");
-    dirData.put(gcsiHelper.getPath(uniqueBucketName2, "foo/bar"),
-                new MethodOutcome(MethodOutcome.Type.RETURNS_TRUE));
+    dirData.put(
+        gcsiHelper.getPath(uniqueBucketName2, "foo/bar"),
+        new MethodOutcome(
+            gcsiHelper.getClass().getSimpleName().equals("HadoopFileSystemIntegrationHelper")
+                ? MethodOutcome.Type.RETURNS_TRUE
+                : MethodOutcome.Type.THROWS_EXCEPTION));
 
     // Call mkdirs() for each path and verify the expected behavior.
     for (URI path : dirData.keySet()) {
@@ -842,7 +843,7 @@ public class GoogleCloudStorageFileSystemIntegrationTest {
         boolean result = gcsiHelper.mkdirs(path);
         if (result) {
           assertWithMessage(
-                  "Unexpected result for path: %s : expected %s, actually returned true.",
+                  "Unexpected result for path: '%s' : expected %s, actually returned true.",
                   path, expectedOutcome)
               .that(expectedOutcome.getType())
               .isEqualTo(MethodOutcome.Type.RETURNS_TRUE);
@@ -850,20 +851,20 @@ public class GoogleCloudStorageFileSystemIntegrationTest {
           // Assert that all of the sub-dirs have been created.
           List<URI> subDirPaths = getSubDirPaths(path);
           for (URI subDirPath : subDirPaths) {
-            assertWithMessage("Sub-path %s of path %s not found or not a dir", subDirPath, path)
+            assertWithMessage("Sub-path '%s' of path '%s' not found or not a dir", subDirPath, path)
                 .that(gcsiHelper.exists(subDirPath) && gcsiHelper.isDirectory(subDirPath))
                 .isTrue();
           }
         } else {
           assertWithMessage(
-                  "Unexpected result for path: %s : expected %s, actually returned false.",
+                  "Unexpected result for path: '%s' : expected '%s', actually returned false.",
                   path, expectedOutcome)
               .that(expectedOutcome.getType())
               .isEqualTo(MethodOutcome.Type.RETURNS_FALSE);
         }
       } catch (Exception e) {
         assertWithMessage(
-                "Unexpected result for path: %s : expected %s, actually threw exception %s.",
+                "Unexpected result for path: '%s' : expected '%s', actually threw exception %s.",
                 path, expectedOutcome, Throwables.getStackTraceAsString(e))
             .that(expectedOutcome.getType())
             .isEqualTo(MethodOutcome.Type.THROWS_EXCEPTION);

